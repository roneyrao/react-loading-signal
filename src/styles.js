export default from './styles/FlexLoading.css';
