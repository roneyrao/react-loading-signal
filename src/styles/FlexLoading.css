.loading {
	width:100%;
	top:50%;
	transform: translateY(-50%);
	z-index:999;
	visibility:hidden;
	text-align:center;
	word-wrap:normal;
	white-space:nowrap;
}
.masked{
	top:0;
	right:0;
	bottom:0;
	left:0;
	z-index:9999;
	display:flex;
	justify-content:center;
	align-items:center;
	transform: none;
}
.masked:before{
	content:'';
	top:0;
	right:0;
	bottom:0;
	left:0;
	background:rgba(150, 150, 150, 0.4);
	position: absolute;
}
.masked.loading>* {
	transform: none;
	z-index: 1;
}

.local{
	composes:loading;
	position:absolute;
}
.localMasked{
	composes:local masked;
}

.global{
	composes:loading;
	position:fixed;
}
.globalMasked{
	composes:global masked;
}

.messageList{
	margin-top:1em;
	line-height:2em;
}
/*theme 'blobs'*/
.blobs{
    width: 100%;
	height:1em;
	background:url(./images/blobs.gif) no-repeat center center;
	background-size:auto 100%;
}
/*theme 'spinning'*/
.spinning{
	min-width:3em;
    padding: 4.5em 1.5em 1.5em;
    border-radius: 0.8em;
	background:rgba(0,0,0,.7) url(./images/spinning.gif) no-repeat center 1.5em;
	background-size:auto 3em;
	color: white;
}





