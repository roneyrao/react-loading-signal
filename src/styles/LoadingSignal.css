.loading {
  top: 50%;
  left: 50%;
  z-index: 999;
  visibility: hidden;
  height: 0;
  text-align:center;
}
.loading > * {
  display: inline-block;
  transform: translate(-50%, -50%);
}
.masked {
  top: 0;
  right: 0;
  bottom: 0;
  left: 0;
  z-index: 9999;
  display: flex;
  justify-content: center;
  align-items: center;
  transform: none;
}
.masked:before {
  content: '';
  top: 0;
  right: 0;
  bottom: 0;
  left: 0;
  background: rgba(150, 150, 150, 0.4);
  position: absolute;
}
.masked.loading > * {
  transform: none;
  z-index: 1;
}

.local {
  composes: loading;
  position: absolute;
}
.localMasked {
  composes: local masked;
}

.global {
  composes: loading;
  position: fixed;
}
.globalMasked {
  composes: global masked;
}

.messageList {
  line-height: 1em;
  list-style: none;
  margin:0;
  padding:0;
}
.messageList > li {
  padding-top: 1em;
}
/*theme 'blobs'*/
.blobs {
  width: 12em;
  height: 1em;
  background: url(./images/blobs.gif) no-repeat center center;
  background-size: auto 100%;
}
/*theme 'spinning'*/
.spinning {
  min-width: 3em;
  max-width: 50%;
  padding: 4.5em 1.5em 1.5em;
  border-radius: 0.8em;
  background: rgba(0, 0, 0, 0.7) url(./images/spinning.gif) no-repeat center 1.5em;
  background-size: auto 3em;
  color: white;
  text-align: center;
  word-wrap: break-word;
}
/*theme 'progress'*/
.progress {
}
.progressBar {
  height: 5px;
  border: 1px solid #057927;
}
.progressBar > div{
  width: 0%;
  height: 100%;
  background: rgb(5, 121, 39);
}
