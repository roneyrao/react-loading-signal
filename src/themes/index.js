export Blobs from './blobs';
export Spinning from './spinning';
export Progress from './progress';
