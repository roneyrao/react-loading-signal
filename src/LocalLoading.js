import React from 'react';
import {
	render,
	findDOMNode,
	unmountComponentAtNode as unmount
} from 'react-dom';
import styles from './styles';
//import styles from './styles/FlexLoading.css';
import PropTypes from 'prop-types';
import {
	Blobs
} from './themes';
import {
	LoadingEvent,
	EVENT_LOCAL_HADNDLED
} from './LoadingEvent';

/*
 * Local loading state is managed by caller;
 */
export default class LocalLoading extends React.Component {
	static propTypes = {
		active: PropTypes.bool.isRequired,
		masked: PropTypes.bool,
		message: PropTypes.any,
		button: PropTypes.instanceOf(Element), //if supplied, 'disabled' of property and class are set;
		// where loading is placed;
		container: PropTypes.oneOfType([PropTypes.instanceOf(Element), PropTypes.oneOf([false, undefined])]),
		theme:PropTypes.func
	}
	static defaultProps = {
		active: false,
		masked: false,
		message: '',
		theme: Blobs
	}
	createLoadingBox(container, masked) {
		this.box = document.createElement('div');
		//console.error('masked', masked);
		this.box.setAttribute('class', masked ? styles.localMasked : styles.local);
		if (!masked) {
			const styles = window.getComputedStyle?window.getComputedStyle(container):container.currentStyle;
			if (styles.position == 'static') {
				container.style.position = 'relative';
			}
		}
		container.appendChild(this.box);
	}
	renderLoading(Theme, message) {
		//console.log('Theme', Theme);
		render(<Theme message={message} />, this.box);
	}
	toggleDisplay(active, button) {
		//when button changed, restore previous button state;
		if(button != this.props.button && this.props.button && this.props.active){
			this.props.button.classList.remove('disabled');
			this.props.button.removeAttribute('disabled');
		}
		if (active) {
			if(this.box){
				this.box.style.visibility = 'visible';
				LoadingEvent.emit(EVENT_LOCAL_HADNDLED);//notify global loading to bypass this one;
			}
			if (button) {
				button.classList.add('disabled');
				button.setAttribute('disabled', 'disabled');
			}
		} else {
			if(this.box){
				this.box.style.visibility = 'hidden';
			}
			if (button) {
				button.classList.remove('disabled');
				button.removeAttribute('disabled');
			}
		}
	}
	componentWillReceiveProps({
		active,
		button,
		container,
		message,
		masked,
		theme
	}) {
		//initial container
		if (container !== this.props.container) {
			if (this.box) {
				this.destroy();
			}
			if (container === undefined) {
				container = findDOMNode(this).parentNode; //eslint-disable-line react/no-find-dom-node
			}
			if(container){
				this.createLoadingBox(container, masked);
			}
		}else{
			if(this.box&&masked!=this.props.masked){
				this.box.setAttribute('class', masked ? styles.localMasked : styles.local);
			}
		}

		//render box content
		if (this.box && (message != this.props.message || theme != this.props.theme)) {
			this.renderLoading(theme, message);
		}

		//show/hide box;
		if (active != this.props.active || button != this.props.button) {
			this.toggleDisplay(active, button);
		}
	}
	componentDidMount() {
		let {
			active,
			masked,
			button,
			container,
			message,
			theme
		} = this.props;
		//console.log('didMount props', this.props);
		//initial container
		if (container === undefined) {
			container = findDOMNode(this).parentNode; //eslint-disable-line react/no-find-dom-node
			//console.log('parent', container);
		}
		if (container) {
			this.createLoadingBox(container, masked);
			this.renderLoading(theme, message);
		}

		this.toggleDisplay(active, button);
	}
	componentWillUnmount() {
		if (this.box) {
			this.destroy();
		}
	}
	destroy(){
		//clear
		unmount(this.box);
		if (this.box.parentNode) {
			this.box.parentNode.removeChild(this.box);
		}
	}
	//nothing to render; done manually into container;
	render() {
		return <i />; //necessary for findDOMNode;
	}
}
