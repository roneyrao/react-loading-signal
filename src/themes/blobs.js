import React from 'react';
import styles from '../styles';
export default function LoadingBlobs(){
	return (<div className={styles.blobs}/>);
}
