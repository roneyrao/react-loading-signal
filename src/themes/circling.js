// @flow
import React from 'react';
import styles from '../styles';

export default function LoadingSpinning({ message }: { message?: React.ChildrenArray<mixed> }) {
  return <div className={styles.circling}>{message}</div>;
}
LoadingSpinning.defaultProps = {
  message: 'loading...',
};
