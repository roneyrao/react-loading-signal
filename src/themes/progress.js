// @flow
import React from 'react';
import styles from '../styles';

type Caption = (number) => string;
type Props = {
  message?: {
    progress: number,
    caption: Caption,
  },
};


function caption(progress) {
  return `current progress: ${(progress * 100).toPrecision(4)}%`;
}

export default function LoadingProgress({ message }: Props) {
  return (
    <div className={styles.progress}>
      <div>{(message.caption || caption)(message.progress)}</div>
      <div className={styles.progressBar}>
        <div style={{ width: `${message.progress * 100}%` }} />
      </div>
    </div>
  );
}
LoadingProgress.defaultProps = {
  message: {
    progress: 0,
  },
};
