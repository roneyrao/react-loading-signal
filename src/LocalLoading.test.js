import React from 'react';
import { mount } from 'enzyme';

import LocalLoading from './LocalLoading';
import styles from './styles';
import * as Themes from './themes';

function getCtnr(wrapper) {
  return wrapper.getDOMNode().parentNode.querySelector(`.${styles.loading}`);
}
describe('initial props', () => {
  test('render properly with default props', () => {
    const ctnr = document.createElement('div');
    ctnr.style.position = 'static'; // no default value as in browser;
    const wrapper = mount(<LocalLoading active={false} />, { attachTo: ctnr });
    const loadingBox = getCtnr(wrapper);

    expect(ctnr.style.position).toBe('relative');
    expect(loadingBox).not.toBeNull();
    expect(loadingBox.style.visibility).toBe('hidden');
    expect(ctnr.outerHTML).toMatchSnapshot();
  });
  test('shown, no error', () => {
    const wrapper = mount(<LocalLoading active />);
    expect(getCtnr(wrapper).style.visibility).toBe('visible');
  });
  test('shown, Id is flagged', () => {
    const id = {};
    const wrapper = mount(<LocalLoading active={id} />);
    expect(getCtnr(wrapper).style.visibility).toBe('visible');
    expect(id._canceled).toBe(true);
  });
  test('masked', () => {
    const wrapper = mount(<LocalLoading active={false} masked />);
    expect(getCtnr(wrapper).classList.contains(styles.masked)).toBe(true);
  });
  test('supply message and theme', () => {
    // Blobs won't show message;
    const msg = 'this is a message';
    const wrapper = mount(<LocalLoading active={false} message={msg} theme={Themes.Spinning} />);
    expect(getCtnr(wrapper).querySelector(`.${styles.spinning}`).innerHTML).toBe(msg);
  });
  test('specify button', () => {
    const ctnr = document.createElement('div');
    const box = document.createElement('div');
    const btn = document.createElement('button');
    ctnr.appendChild(btn);
    ctnr.appendChild(box);
    mount(<LocalLoading button={btn} active />, { attachTo: box });
    expect(btn.classList.contains('disabled')).toBe(true);
    expect(btn.getAttribute('disabled')).toBeDefined();
  });
  test('specify container', () => {
    const ctnr = document.createElement('div');
    mount(<LocalLoading active={false} container={ctnr} />);
    expect(ctnr.querySelector(`.${styles.loading}`)).not.toBeNull();
  });
  test("don't show loading by setting container to false", () => {
    const id = {};
    const btn = document.createElement('button');
    const wrapper = mount(<LocalLoading button={btn} active={id} container={false} />);

    expect(getCtnr(wrapper)).toBeNull();
    expect(btn.classList.contains('disabled')).toBe(true);
    expect(id.canceled).toBeUndefined();
  });
});

describe('change props', () => {
  describe('active', () => {
    test('undefined/hidden -> visible', () => {
      const wrapper = mount(<LocalLoading active={false} />);
      const loadingBox = getCtnr(wrapper);
      expect(loadingBox.style.visibility).toBe('hidden');
      wrapper.setProps({ active: true });
      expect(loadingBox.style.visibility).toBe('visible');
    });
    test('visible -> hidden', () => {
      const btn = document.createElement('button');
      const wrapper = mount(<LocalLoading active button={btn} />);
      const loadingBox = getCtnr(wrapper);
      expect(loadingBox.style.visibility).toBe('visible');
      wrapper.setProps({ active: false });
      expect(loadingBox.style.visibility).toBe('hidden');
    });
  });
  describe('masked', () => {
    test('undefined/unmasked -> masked', () => {
      const wrapper = mount(<LocalLoading active={false} />);
      const loadingBox = getCtnr(wrapper);
      expect(loadingBox.classList.contains(styles.masked)).toBe(false);
      wrapper.setProps({ masked: true });
      expect(loadingBox.classList.contains(styles.masked)).toBe(true);
    });
    test('masked -> unmasked', () => {
      const wrapper = mount(<LocalLoading active={false} masked />);
      const loadingBox = getCtnr(wrapper);
      expect(loadingBox.classList.contains(styles.masked)).toBe(true);
      wrapper.setProps({ masked: false });
      expect(loadingBox.classList.contains(styles.masked)).toBe(false);
    });
  });
  describe('message', () => {
    function getHtml(box) {
      return box.querySelector(`.${styles.spinning}`).innerHTML;
    }
    test('undefined/empty -> new message', () => {
      const msg = 'this is a message';
      const wrapper = mount(<LocalLoading theme={Themes.Spinning} active={false} />);
      const box = getCtnr(wrapper);
      expect(getHtml(box)).toBe('loading...');
      wrapper.setProps({ message: msg });
      expect(getHtml(box)).toBe(msg);
    });
    test('with message -> empty', () => {
      const msg = 'this is a message';
      const wrapper = mount(<LocalLoading theme={Themes.Spinning} active={false} message={msg} />);
      const box = getCtnr(wrapper);
      expect(getHtml(box)).toBe(msg);
      wrapper.setProps({ message: '' });
      expect(getHtml(box)).toBe('');
    });
    test('with message -> another message', () => {
      const msg1 = 'this is a message';
      const msg2 = 'this is another message';
      const wrapper = mount(<LocalLoading theme={Themes.Spinning} active={false} message={msg1} />);
      const box = getCtnr(wrapper);
      expect(getHtml(box)).toBe(msg1);
      wrapper.setProps({ message: msg2 });
      expect(getHtml(box)).toBe(msg2);
    });
  });
  describe('button', () => {
    test('undefined/no button -> button', () => {
      const ctnr = document.createElement('div');
      const box = document.createElement('div');
      const btn = document.createElement('button');
      ctnr.appendChild(btn);
      ctnr.appendChild(box);
      const wrapper = mount(<LocalLoading active />, { attachTo: box });
      expect(btn.classList.contains('disabled')).toBe(false);
      wrapper.setProps({ button: btn });
      expect(btn.classList.contains('disabled')).toBe(true);
    });
    test('button -> no button', () => {
      const ctnr = document.createElement('div');
      const box = document.createElement('div');
      const btn = document.createElement('button');
      ctnr.appendChild(btn);
      ctnr.appendChild(box);
      const wrapper = mount(<LocalLoading button={btn} active />, { attachTo: box });
      expect(btn.classList.contains('disabled')).toBe(true);
      wrapper.setProps({ button: null });
      expect(btn.classList.contains('disabled')).toBe(false);
    });
    test('button -> another button', () => {
      const ctnr = document.createElement('div');
      const box = document.createElement('div');
      const btn = document.createElement('button');
      const btn2 = document.createElement('button');
      ctnr.appendChild(btn);
      ctnr.appendChild(btn2);
      ctnr.appendChild(box);
      const wrapper = mount(<LocalLoading button={btn} active />, { attachTo: box });
      expect(btn.classList.contains('disabled')).toBe(true);
      expect(btn2.classList.contains('disabled')).toBe(false);
      wrapper.setProps({ button: btn2 });
      expect(btn.classList.contains('disabled')).toBe(false);
      expect(btn2.classList.contains('disabled')).toBe(true);
    });
  });
  describe('container', () => {
    test('undefined/parentNode -> DOM', () => {
      const ctnr = document.createElement('div');
      const wrapper = mount(<LocalLoading active={false} />);
      expect(getCtnr(wrapper)).not.toBeNull();
      wrapper.setProps({ container: ctnr });
      expect(getCtnr(wrapper)).toBeNull();
      expect(ctnr.querySelector(`.${styles.loading}`)).not.toBeNull();
      expect(ctnr.querySelector(`.${styles.blobs}`)).not.toBeNull();
    });
    test('undefined/parentNode -> false/no container', () => {
      const wrapper = mount(<LocalLoading active={false} />);
      expect(getCtnr(wrapper)).not.toBeNull();
      wrapper.setProps({ container: false });
      expect(getCtnr(wrapper)).toBeNull();
    });
    test('DOM -> undefined/parentNode', () => {
      const ctnr = document.createElement('div');
      const wrapper = mount(<LocalLoading active={false} container={ctnr} />);
      expect(ctnr.querySelector(`.${styles.loading}`)).not.toBeNull();
      wrapper.setProps({ container: undefined });
      expect(ctnr.querySelector(`.${styles.loading}`)).toBeNull();
      expect(getCtnr(wrapper)).not.toBeNull();
    });
    test('DOM -> false/no container', () => {
      const ctnr = document.createElement('div');
      const wrapper = mount(<LocalLoading active={false} container={ctnr} />);
      expect(ctnr.querySelector(`.${styles.loading}`)).not.toBeNull();
      wrapper.setProps({ container: false });
      expect(ctnr.querySelector(`.${styles.loading}`)).toBeNull();
      expect(getCtnr(wrapper)).toBeNull();
    });
    test('false/no container -> DOM', () => {
      const ctnr = document.createElement('div');
      const wrapper = mount(<LocalLoading active={false} container={false} />);
      expect(getCtnr(wrapper)).toBeNull();
      wrapper.setProps({ container: ctnr });
      expect(ctnr.querySelector(`.${styles.loading}`)).not.toBeNull();
    });
    test('false/no container -> undefined/parentNode', () => {
      const wrapper = mount(<LocalLoading active={false} container={false} />);
      expect(getCtnr(wrapper)).toBeNull();
      wrapper.setProps({ container: undefined });
      expect(getCtnr(wrapper)).not.toBeNull();
    });
  });

  describe('theme', () => {
    test('undefined/default theme -> theme', () => {
      const wrapper = mount(<LocalLoading active={false} />);
      const box = getCtnr(wrapper);
      expect(box.querySelector(`.${styles.blobs}`)).not.toBeNull();
      wrapper.setProps({ theme: Themes.Spinning });
      expect(box.querySelector(`.${styles.blobs}`)).toBeNull();
      expect(box.querySelector(`.${styles.spinning}`)).not.toBeNull();
    });
    test('theme -> undefined/default theme', () => {
      const wrapper = mount(<LocalLoading active={false} theme={Themes.Spinning} />);
      const box = getCtnr(wrapper);
      expect(box.querySelector(`.${styles.blobs}`)).toBeNull();
      wrapper.setProps({ theme: null });
      expect(box.querySelector(`.${styles.blobs}`)).not.toBeNull();
      expect(box.querySelector(`.${styles.spinning}`)).toBeNull();
    });
    test('theme -> another theme', () => {
      const wrapper = mount(<LocalLoading active={false} theme={Themes.Spinning} />);
      const box = getCtnr(wrapper);
      expect(box.querySelector(`.${styles.spinning}`)).not.toBeNull();
      wrapper.setProps({ theme: Themes.Circling });
      expect(box.querySelector(`.${styles.spinning}`)).toBeNull();
      expect(box.querySelector(`.${styles.circling}`)).not.toBeNull();
    });
  });
});

test('destroy', () => {
  const wrapper = mount(<LocalLoading />);
  const inst = wrapper.instance();
  const ctnr = getCtnr(wrapper);
  expect(ctnr).not.toBeNull();
  wrapper.unmount();
  expect(inst.box).toBeNull();
  expect(ctnr.querySelector(`.${styles.loading}`)).toBeNull();
});
