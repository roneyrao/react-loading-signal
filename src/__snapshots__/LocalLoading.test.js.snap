// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`initial props render properly with default props 1`] = `"<div style=\\"position: relative;\\"><i></i><div class=\\"LoadingSignal__local LoadingSignal__loading\\" style=\\"visibility: hidden;\\"><div class=\\"LoadingSignal__blobs\\"></div></div></div>"`;
